### PR TITLE
Apple M1 FPWM (MacBook Pro keyboard backlight) driver

### PR DESCRIPTION
**Pull request by pipcet**
I found this very useful during development, since it can be bound to an activity LED and provides visual feedback even when the framebuffer is broken or disabled.

Tested using the sysfs interface:
* enable enables/disables the keyboard backlight
* duty_cycle and period work as expected for setting apparent brightness
* duty_cycle and period work as expected with long periods, so a period value of 1000000000 ns corresponds to a one-second period

This also adds the J293 devicetree (I don't know whether the MacBook Air has a similar keyboard backlight). It does not include the bindings for setting up an LED device backed by the PWM, because I wanted to test the PWM driver without interference from another device.

### CODE REVIEW

File: drivers/pwm/pwm-apple-m1.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	struct resource *res;
+	int ret;
+
+	fpwm = devm_kzalloc(&pdev->dev, sizeof(*fpwm), GFP_KERNEL);
+	if (!fpwm)
+		return -ENOMEM;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!res)
+		return -ENODEV;
```
Comment (alyssarosenzweig) [inline drivers/pwm/pwm-apple-m1.c:104]: Retracted ~~fpwm is leaked here~~ you're fine

File: drivers/pwm/pwm-apple-m1.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+static int fpwm_remove(struct platform_device *pdev)
+{
+	struct fpwm_chip *fpwm = platform_get_drvdata(pdev);
+
+	clk_disable_unprepare(fpwm->clk);
+
+	return 0;
+}
+
+#ifdef CONFIG_OF
```
Comment (alyssarosenzweig) [inline drivers/pwm/pwm-apple-m1.c]: You depend on OF, so this #ifdef might be redundant?

**Comment by alyssarosenzweig** [comment]
Looks good to me. @svenpeter42 please test on macbook air

File: drivers/pwm/pwm-apple-m1.c (hunk 2 above)
Comment (pipcet) [reply drivers/pwm/pwm-apple-m1.c]: So it is. Fixed now.

**Comment by pipcet** [comment]
> Looks good to me. @svenpeter42 please test on macbook air

It might be a good idea to check whether the ADT looks the same first. Also the clock obviously needs to be enabled in m1n1.

**Comment by pipcet** [comment]
Fixed an issue that was reported by the kernel build robot. Also fixed some issues reported by `checkpatch`.

File: drivers/pwm/pwm-apple-m1.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+#include <linux/io.h>
+#include <linux/math64.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/pwm.h>
+#include <linux/slab.h>
+
+#define FPWM_CONTROL		0x00
+#define   CONTROL_UPDATE	   0x4239
```
Comment (pipcet) [inline drivers/pwm/pwm-apple-m1.c:36]: As far as I'm concerned, that's just a magic constant, derived at experimentally. Should there be a comment explaining that?

**Comment by pipcet** [comment]
@amworsley tested this on a MacBook Air and it seems to be working.

**Comment by zzywysm** [comment]
Did you intend to close this pull request?

**Pull request closed without merging**